core/src/main/java/org/elasticsearch/index/query/GeoShapeQueryBuilder.java

```diff
@@ -589,11 +621,21 @@ public String getWriteableName() {
     }
 
     @Override
-    protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
-        if (this.shape == null) {
-            GetRequest getRequest = new GetRequest(indexedShapeIndex, indexedShapeType, indexedShapeId);
-            ShapeBuilder shape = fetch(queryShardContext.getClient(), getRequest, indexedShapePath);
-            return new GeoShapeQueryBuilder(this.fieldName, shape).relation(relation).strategy(strategy);
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        if (supplier != null) {
+            return supplier.get() == null ? this : new GeoShapeQueryBuilder(this.fieldName, supplier.get()).relation(relation).strategy
+                (strategy);
+        } else if (this.shape == null) {
+            AtomicReference<ShapeBuilder> supplier = new AtomicReference<>();
```

Comment:
I wonder if `AtomicReference` is needed here given that we write from a single thread. It's just a visibility problem hence `SetOnce` would be a good fit, which has also the advantage of checking that we do set it only once (thanks for the suggestion!)